Enhance P036 - Framed OLED to support more sizes and customized header

uwekaditz opened the pull request:
[P036] NEW more OLED sizes (128x32, 64x48) added to the original 128x64 size
[P036] NEW Display button can be inverted
[P036] NEW Content of header is adjustable, also the alternating function
[P036] CHG Parameters sorted

uwekaditz committed:
Update _P036_FrameOLED.ino

uwekaditz committed:
Update _P036_FrameOLED

[P036] NEW more OLED sizes (128x32, 64x48) added to the original 128x64 size
[P036] NEW Display button can be inverted
[P036] NEW Content of header is adjustable, also the alternating function
[P036] CHG Parameters sorted

TD-er commented:
I changed the title to be a bit more descriptive.

Can you maybe also have a look at the bottom line, where the dots are being displayed.
There have been some requests for having a more useful display there instead of page numbering.
Maybe there can be some parts shown which now are also available for the header.
And pagination can also be shown like `[2/5]` or something like that.

uwekaditz commented:
On the smaller displays the bottom line is not shown.
Yesterday I ordered the 128x64 display, but the delivery is not before December 2019, China!!
I can add the page numbers to the customizable header.

Sasch600xt commented:
Hello :)
Thank you for working on it :)
Would be also great if i could send content to the display by using http or mqtt :)

Have a great day

fly74 commented:
Nice to hear. In which build will be a test possible?

uwekaditz commented:
@Sasch600xt,
Why do you need the content of the display broadcasted by http or mqtt?
The content of the display is mainly taken from other sensors, and they have the opportunity to send the measured values to the defined controllers, like http, them or mqtt.

TD-er commented:
> Why do you need the content of the display broadcasted by http or mqtt?

No, the other way around.
As if you would send a command to display something on screen.

Sasch600xt commented:
TD-er is right.

it would be great to send some content from external source to the display by MQTT or Http.

uwekaditz commented:
I'm sorry, I'm a more hardware related programmer. If anybody has some ideas how to access a http or mqtt message within the 1 second call of the plug in, please let me know.

TD-er commented:
@uwekaditz I think it would be best to have a command added.
Comands are processed in the `PLUGIN_WRITE` section.
There is already one command for P036: `OLEDFRAMEDCMD`
But that one is not really using the best command parser functions, so I guess we can also split it into several fixes. One for the enhancements you made now and one for the extending of commands for this plugin.

I will make a test build for this PR, so others can also test it.

uwekaditz committed:
Merge remote-tracking branch 'upstream/mega' into PullRequest

TD-er commented:
I created a test build, based on this PR: [ESPEasy_mega-20191104-3-PR_2706.zip](https://www.dropbox.com/s/5n67ba3ewv93l5c/ESPEasy_mega-20191104-3-PR_2706.zip?dl=0)

I did not even test it myself, so test only on units that are easy to reach for serial flash.

Edit:
The build started before the upstream was merged into this PR.

uwekaditz committed:
[P036] Added pagination to customized header

[P036] Added pagination to customized header
[P036] Time can be always selected in customized header, even if the time is diplayed by default for 128x68

uwekaditz committed:
Update _P036_FrameOLED.ino

[P036] BUG Rotated image did not work for displays with height less then 64 pix

uwekaditz commented:
Just changed P036_FrameOLED!
NEW: Added pagination to customized header
BUG Rotated image did not work for displays with height less then 64 pix
CHG: Time can be always selected in customized header, even if the time is displayed by default for display 128x64 below the header

uwekaditz commented:
**@mqtt import**
the plugin P037_MQTTImport ("Generic - MQTT Import") can be already used to read data from the MQTT import input stream and save the values
the saved value can be displayed if one of the content lines is defined as [_{mqtt import name}_#Value1] (**_Note:_** _{mqtt import name}_ is the given task name of the device "Generic - MQTT Import")
**@http import**
why not creating a plugin similar to _P037_MQTTImport ("Generic - MQTT Import") just using a HTTP import

TD-er commented:
> why not creating a plugin similar to _P037_MQTTImport ("Generic - MQTT Import") just using a HTTP import

Well we have a lot of ways to enter commands, so the usual way is to implement a command for a plugin, which then can be sent via a lot of ways.
Since I also typed a very elaborate reply on commands in another issue this morning, I will just link to it: https://github.com/letscontrolit/ESPEasy/issues/2712#issuecomment-549294948

In short:
A command can be more than just about displaying a value. For example texts cannot be sent via MQTT import, since that one only results a floating point value.
Also a command can do more, like clearing a line, showing a text or switching to a specific page.

fly74 commented:
what means the double zero in the headline?

[8266.zip](https://github.com/letscontrolit/ESPEasy/files/3805602/8266.zip)

uwekaditz commented:
Is this already the changed plugin?
**If yes**, what did you set for the header content and what is the size of your display? (see picture)
**If no**, the header displays alternating WiFi.SSID (if connected) and %sysname%.
![header](https://user-images.githubusercontent.com/55511593/68150399-f336fb00-ff3f-11e9-9e0f-4c906eeb93eb.JPG)

fly74 commented:
![2019-11-04 20_20_57-ESPOLEDSW und 1 weitere Seite - Profil 1 - Microsoft Edge](https://user-images.githubusercontent.com/38754941/68150748-a69fef80-ff40-11e9-8178-7badcb2f1f83.png)

Time and Pagenumbers can't be selected:

![2019-11-04 20_23_34-ESPOLEDSW und 1 weitere Seite - Profil 1 - Microsoft Edge](https://user-images.githubusercontent.com/38754941/68150941-09918680-ff41-11e9-9e3c-b8be30b63aee.png)

uwekaditz commented:
And what is the content of %sysname% in **_Tools->System Variables_**

fly74 commented:
%sysname% | ESPOLEDSW | ESPOLEDSW
-- | -- | --

fly74 commented:
I've it. It's the minute from actual time. But why there.

uwekaditz commented:
I found the problem!
For the 128x64 and 128x32 displays the wifi bars should be on the right side as it was before my changes.
Now is is on the left and overwrites the actual time.
I'll correct it in a few minutes.

uwekaditz committed:
[P036] BUG Wifi bars for 128x64 and 128x32 displays must be on right side

Sasch600xt commented:
Wow !!! Thank you so much ! I will test it tomorrow.
As long the MQTT import works i do not need a http solution.
Maybe other users would need it.
I was just giving options "Mqtt or/and http"

But MQTT for communication to the display is brilliant !

Have a great day

uwekaditz commented:
I'm working on another feature:

- scrolling long lines which do not fit to the OLED screen width
- scrolling will be from right to left
- each line can scroll independently (depending on its width)
- before and after scrolling there will be a certain waiting time

This feature will be ready by the end of this week.

TD-er commented:
Hopefully you will also allow a "no scroll" option?
This plugin already takes quite some time in blocking code execution. See the timing stats page.
My ideal is to have all timers (also max. values) below 100 msec. max.

uwekaditz commented:
Of course there will be a 'no scroll' option.
The new line scrolling is done in the PLUGIN_TEN_PER_SECOND call.
The already existing frame scrolling is done in the adjustable timer call PLUGIN_READ.
Attached is the timing measured with a Wemos 4M1M (80Mhz, D1R1).
![OLED-Timing](https://user-images.githubusercontent.com/55511593/68252078-d1637400-0024-11ea-8451-b03c3e6f12af.JPG)

TD-er commented:
700 msec is way too long, so I hope you can optimize that a bit.
Updating it in the 10/sec call does sound OK, as long as it does not hurt the number of calls to the other 10/sec calls too much.
Those should be called as close as possible to the 10/sec.

uwekaditz commented:
I didn't touch the already existing frame scrolling so far.
This srolling is done by a stupid and blocking **_for_** loop (see attached code snippet)
[display_scroll.txt](https://github.com/letscontrolit/ESPEasy/files/3811634/display_scroll.txt)

TD-er commented:
I know the existing code is far from optimal, so that's why I am happy you're having a look at it.
Just try to keep in mind that shorter blocking loops is better.
I don't mind if the CPU-load is 100%, as long as all scheduled tasks can be run in time.

Maybe you can even optimize the scrolling to skip an update if your 10/sec call missed a call.
Such an optimization does make the scroll speed look more natural and it does help the scheduler to meet its deadlines, regardless of the number of tasks active on the node.

uwekaditz commented:
I made a short video of the line scrolling function.
[20191105_235529_1.zip](https://github.com/letscontrolit/ESPEasy/files/3811667/20191105_235529_1.zip)

Sasch600xt commented:
Hello :)

will there be also an kinda "simple mode" ? So no scrolling, no page changing feature at all.
So i am talking about having one line with big letters, 2 Lines with medium letters or 4 lines with small letters. So i would love to send content with MQTT to the display. (like you explained already)

A killer feature would be to change the line settings (1 line, 2 line or 4 line) by MQTT. So i could send a big "OK" to the display as one big line. After that i switch to "4 line mode" and send content for 4 lines.

But maybe this is something fo much much later :)
But if you can, please think about an really "simple mode" without any animation just to use it as a "normal" display with 1,2 or 4 lines.

An thank you soooo much for your help with this plugin :)
You guys rock :)

uwekaditz commented:
Hi,thus simple Mode ist already there.You just need to set the correct parameters:lines per pagescrolling speed = instantheader = header ( alternating)If you fill only as many lines as you defined per page there is also no page changing.RegardsVon meinem Samsung Galaxy S8 gesendet.
-------- Ursprüngliche Nachricht --------Von: Sasch600xt <notifications@github.com> Datum: 06.11.19  00:13  (GMT+01:00) An: letscontrolit/ESPEasy <ESPEasy@noreply.github.com> Cc: uwekaditz <mueller.dresden@t-online.de>, Mention <mention@noreply.github.com> Betreff: Re: [letscontrolit/ESPEasy] Enhance P036 - Framed OLED to support more sizes and customized header (#2706) Hello :)
will there be also an kinda "simple mode" So no scrolling, no page changing feature at all.
So i am talking about having one line with big letters, 2 Lines with medium letters or 4 lines with small letters. So i would love to send content with MQTT to the display. (like you explained already)
A killer feature would be to change the line settings (1 line, 2 line or 4 line) by MQTT. So i could send a big "OK" to the display as one big line. After that i switch to "4 line mode" and send content for 4 lines.
But maybe this is something fo much much later :)
But if you can, please think about an really "simple mode" without any animation just to use it as a "normal" display with 1,2 or 4 lines.
An thank you soooo much for your help with this plugin :)
You guys rock :)

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/letscontrolit/ESPEasy/pull/2706?email_source=notifications\u0026email_token=ANHQUKJDGIPQECVSJPV3UDTQSH42FA5CNFSM4JII6D32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDEVPIA#issuecomment-550066080",
"url": "https://github.com/letscontrolit/ESPEasy/pull/2706?email_source=notifications\u0026email_token=ANHQUKJDGIPQECVSJPV3UDTQSH42FA5CNFSM4JII6D32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDEVPIA#issuecomment-550066080",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

TD-er commented:
What he meant was a special command (or a parameter for a command) to not only set some text to a specific line, but also to switch between modes.
For example:
```
oledwrite,"text",1   // Write "text" on line 1
oledwrite,"text",1,2   // Write "text" on line 1, 2 line mode
```
Not that this should be the command structure (or command name), just as an example.

uwekaditz committed:
Merge remote-tracking branch 'upstream/mega' into PullRequest

uwekaditz committed:
[P036] Time for PLUGIN_READ reduced

CHG: Page scrolling based on Timer (PLUGIN_TIMER_IN) to reduce time for PLUGIN_READ from 700ms to less than 80ms (not as smooth as before but non-blocking!)

uwekaditz commented:
Yesterday I changed the [P036] again:
- Page scrolling based on Timer (PLUGIN_TIMER_IN) to reduce time for PLUGIN_READ from 700ms to less than 80ms (not as smooth as before but non-blocking!)

The next task will be the command OLEDFRAMEDCMD, but I got some questions:
1. Is there a limitation for static variables in the plug in?
2. Can I save and open another custom setting file using the function LoadCustomTaskSettings(taskIndex,...) to avoid an incompatibility to the original data already saved in such a custom setting file?
3. Is there another way to save a bigger structure (strings, fonts and height)?

fly74 commented:
where can I get the new binary?

uwekaditz commented:
It has to be merged to the next daily build by TD-er.
On the other hand you can check out my repository and compile it by yourself.

TD-er commented:
I will check it this evening and if it doesn't burn down the house, I will merge it ;)

TD-er commented:
OK, I can put away the fire extinguisher and take off my hat and bow for this plugin update.


Description | Function | #calls | call/sec | min (ms) | Avg (ms) | max (ms)
-- | -- | -- | -- | -- | -- | --
P_36_Display - OLED SSD1306/SH1106 Framed | READ | 48 | 0.10 | 43.172 | 56.612 | 73.348
P_36_Display - OLED SSD1306/SH1106 Framed | ONCE_A_SECOND | 461 | 1.00 | 9.858 | 11.316 | 14.972
P_36_Display - OLED SSD1306/SH1106 Framed | TEN_PER_SECOND | 4576 | 9.92 | 0.020 | 0.490 | 21.108
P_36_Display - OLED SSD1306/SH1106 Framed | EVENT_OUT | 48 | 0.10 | 0.015 | 0.020 | 0.035
P_36_Display - OLED SSD1306/SH1106 Framed | UDP_IN | 8 | 0.02 | 0.003 | 0.008 | 0.011
P_36_Display - OLED SSD1306/SH1106 Framed | FIFTY_PER_SECOND | 22442 | 48.66 | 0.007 | 0.013 | 0.050


This is with "very slow" scrolling.

I tested with 1.3" SH1106 and 0.96" SSD1306.

Just a few minor issues, which can also be fixed after this.
- Switch frequency between alternating texts on the top line is a bit too fast.
- Scrolling long lines may leave some uncleared lines at the end when a short line is shown afterward.
- On one node with pre-configured OLED Framed controller the plugin was disabled after boot. Enabling and save and it was working. Perhaps a wrong config parameter without checks for correct config?

uwekaditz commented:
1. The header alternating is done once per second. Any other idea?
2. Do the uncleared lines happen with the SH1106 controller? If yes, it is quite clear because this controller is using a 132x64 pixel field and the plug in handles only 128x64, I didn't  change that from the original source code, because I have nothing to test.
3. I didn't change the original config parameters. I used my 64x32 display also with the original plug in and all parameter settings were as they were before after using my updated code, strange!

TD-er commented:
1. Alternating frequency of half the update interval of the pages would be OK I guess. Or minimum of 1x in 2 seconds? Reading of the IP or SSID is a bit hard now, with this switching speed.
2. The not cleared lines do indeed happen on the SH1106 display.
3. You added new parameters and maybe those cannot handle "impossible" values? The existing config does not guarantee what value the unused bytes will have.

fly74 commented:
Display-Timout ist not working (always on) with the last change Release mega-20191113:

uwekaditz (6):
      Update _P036_FrameOLED.ino
      Update _P036_FrameOLED
      [P036] Added pagination to customized header
      Update _P036_FrameOLED.ino
      [P036] BUG Wifi bars for 128x64 and 128x32 displays must be on right side
      [P036] Time for PLUGIN_READ reduced

TD-er commented:
> Display-Timout ist not working (always on) with the last change Release mega-20191113:

Was it working with the older plugin? (I never used it, if it was even present)

fly74 commented:
![2019-11-13 10_47_11](https://user-images.githubusercontent.com/38754941/68752145-11b58a00-0603-11ea-9435-9f14f9a4e464.png)

after the time the display goes off and can be reactivated over the button

uwekaditz commented:
May be you have to check the (new!) invert functionality of the display button.

 

In my application the display button is connected to a photo resistor and it works fine (display is off if it is dark).

fly74 commented:
thats it. thx.

fly74 commented:
is it possible to have a mode without healine an pager? only to have big letter e.g.
temperatures with 4 digit (-11,33°C) over the full display

uwekaditz commented:
So far the header is always visible (containing WiFi bars, content and time (only for 128pix wide displays).

 

I can add another check box to hide the header in the next release.

fly74 commented:
I think hiding is not enough, if the letter not going bigger.

TD-er commented:
The biggest font we have now embedded is the one for 1 line. If the header(s) are removed, that font still will not cover the entire display height.

Maybe we should add a new issue with feature/change requests for this plugin?

uwekaditz commented:
Following fonts are available from the esp8266-oled-ssd1306 library:

 

ArialMT_Plain_24    Height:28 (used in 128x64 and 64x48 - 1 line)

ArialMT_Plain_16    Height:19 (used in 128x64 - 2 lines)

Dialog_plain_12     Height:15 (used in 128x64 - 3 lines; 128x32 – 1 line; 64x48 - 2 lines) 

ArialMT_Plain_10    Height:13 (used for all other settings)

TD-er commented:
Please note that I generated these fonts myself, so there are others possible, but they do add quite a lot to the binary size.

fly74 commented:
If we can do it with a feature request, so the non framed version could be removed. Would be fine.

uwekaditz commented:
I found a web page to create customized font data

http://oleddisplay.squix.ch/#/home

fly74 commented:
A small font with only big letters with -+ 0-9 ,: would by fine for big sensor data. Maybe not so much consuming of space.

________________________________
Von: uwekaditz <notifications@github.com>
Gesendet: Wednesday, November 13, 2019 7:19:52 PM
An: letscontrolit/ESPEasy <ESPEasy@noreply.github.com>
Cc: fly74 <fly74@live.de>; Comment <comment@noreply.github.com>
Betreff: Re: [letscontrolit/ESPEasy] Enhance P036 - Framed OLED to support more sizes and customized header (#2706)

I found a web page to create customized font data

http://oleddisplay.squix.ch/#/home


—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fletscontrolit%2FESPEasy%2Fpull%2F2706%3Femail_source%3Dnotifications%26email_token%3DAJHVU7OJH3OI2QO6WR6PW4DQTRAMRA5CNFSM4JII6D32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED7EJSA%23issuecomment-553534664&data=02%7C01%7C%7Cb0f2d1a3ba654a0aaff108d768661498%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637092659942116356&sdata=AowyycnOc310j7a6QYI1rCxCGX8Ri2QcZrBbqAjxkMM%3D&reserved=0>, or unsubscribe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJHVU7KRA4P725YZYMZ7FW3QTRAMRANCNFSM4JII6D3Q&data=02%7C01%7C%7Cb0f2d1a3ba654a0aaff108d768661498%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637092659942126355&sdata=SzzEtvju9ogB%2BagWAIekUlrEo5m504EZ3cNvvNun1wo%3D&reserved=0>.

TD-er commented:
Let's start an issue for these requests, instead if discussing it in the comment section of an already merged PR.

Sasch600xt commented:
Hello all :)

thank you for the great work already :)
I am playing around with the MQTT Message send to the Display.

It works great already for numeric values.
Will there be in future a solution to send text to the display ?

Thank you so much :)

TD-er commented:
> Will there be in future a solution to send text to the display ?

Yep, very likely there will be.
But implementing a "string" value type does require a lot of work.
Currently there is 1 plugin (highly experimental status) which does have a string type value.
It is mainly to see what we will run into when working with non-numerical value types.

Sasch600xt commented:
ok.....thank you for your answere. so i just wait a little longer :)

TD-er commented:
Adding a command to show text on a display is already present in the new code (if my memory serves me well), but sending it via MQTT is a different story.

Sasch600xt commented:
i use ESP_Easy_mega-20191123_test_beta_ESP8266_4M1M right now.

Can i somehow bring text to the display already ? by http for example ?

Also i try to show some values at the display like:
[Encoder#Encoder_Counter] / [Encoder-Switch#State]

The result at the display would be something like "10 / 0"
But i cant type all line in because it does not allow so many charakters.
it allowes me something like this: [Encoder#Encoder_Counter] / [Enco

Thank you :)

TD-er commented:
The documentation is not yet updated, but purely based on the source I would say you should be able to give commands like this:

```
oledframedcmd,<linenr>,"text"
```
So something like this:
```
oledframedcmd,8,"This is the 8th line"
```
Or as html GET command: http://192.168.1.75/tools?cmd=oledframedcmd%2C8%2C%22This+is+the+8th+line%22

The line nr should correspond with the line numbers on the task configuration.
It does not over write the text in the settings, so after reboot it will start with the programmed text for that line.

Sasch600xt commented:
YAY ! That worked very well  with http://192.168.1.75/tools?cmd=oledframedcmd%2C8%2C%22This+is+the+8th+line%22

I did edit my post i guess after you answered. Did you read the part about the lenght of line i can type in the display plugin ?

TD-er commented:
Just type in your command in the command field on the Tools page and you'll see how you have to escape the special characters to match [task#value] like parts and spaces.  (nope, had not read your edit)

Sasch600xt commented:
i am talking about the limit of characters.

i try to type the following line:
[Encoder#Encoder_Counter] / [Encoder-Switch#State]

But look at the Screenshot, this is the maximum i can type into the field.

![Screenshot 2019-11-24 23 33 03](https://user-images.githubusercontent.com/25459028/69502769-47b41180-0f13-11ea-83d9-aad7228d2a5c.png)

TD-er commented:
Then you can rename the task name or variable name to make it shorter.
There is a pull request as a work-in-progress to increase this limit for the lines, but like I said it is a work-in-progress.

Sasch600xt commented:
ok....good idea for a work around !

i have time, but wonderful features in the new plugin already !
Thank you very much for this work to all of you ! :)
